custom/onse/tasks.py

```diff
@@ -83,9 +85,61 @@ def update_facility_cases_from_dhis2_data_elements(
         otherwise they are emailed.
 
     """
+    _update_facility_cases_from_dhis2_data_elements(period, print_notifications)
+
+
+@task
+def _update_facility_cases_from_dhis2_data_elements(period, print_notifications, retry_attempt=0):
     if not domain_exists(DOMAIN):
         return
     dhis2_server = get_dhis2_server(print_notifications)
+
+    if _server_ready(dhis2_server):
+        _execute_update_facility_cases_from_dhis2_data_elements(dhis2_server, period, print_notifications)
+    else:
+        retry_attempt += 1
+        if retry_attempt <= MAX_RETRY_ATTEMPTS:
+            retry_in_n_days = 2 ** (retry_attempt - 1)
+            retry_date = datetime.utcnow() + relativedelta(days=retry_in_n_days)
+
+            print(f'Importing ONSE ISS facility cases from DHIS2 failed: server unreachable. '
+                  f'Retrying on {retry_date.date()}')
+
+            schedule_execution(
+                _update_facility_cases_from_dhis2_data_elements,
+                [period, print_notifications, retry_attempt],
+                retry_date
+            )
+        else:
+            print('Importing ONSE ISS facility cases from DHIS2 failed: server unreachable. '
+                  'No more retry attempts')
+
+
+def schedule_execution(callback_task, args: list, on_date: datetime):
+    always_eager = hasattr(settings, "CELERY_TASK_ALWAYS_EAGER") and settings.CELERY_TASK_ALWAYS_EAGER
+    settings.CELERY_TASK_ALWAYS_EAGER = False
+    callback_task.apply_async(tuple(args), eta=on_date)
+    settings.CELERY_TASK_ALWAYS_EAGER = always_eager
```

Comment:
Wasn't my first choice, but apparently with `CELERY_TASK_ALWAYS_EAGER=True` even the `apply_async` tasks gets executed immediately.

I found an example in the codebase [here](https://github.com/dimagi/commcare-hq/blob/4a64a8d722724d80ddebbceff65e9d7b1290468d/corehq/apps/userreports/reports/builder/forms.py#L1098).

Reply:
> I found an example in the codebase here.

Hmmm. That does look a bit weird.

What if you pass a timedelta instead of a datetime? e.g. [This SO answer](https://stackoverflow.com/a/45084288).

Reply:
@kaapstorm I did come across that comment, _but_ [this](https://docs.celeryproject.org/en/3.1/configuration.html#celery-always-eager) was basically the reason I decided to rather fiddle with the settings. (tl;dr : it's not actually being sent to the queue if `CELERY_TASK_ALWAYS_EAGER=True`; the setting has been [renamed](https://stackoverflow.com/a/44319939) since celery 4). [This](https://stackoverflow.com/questions/45083908/when-i-use-django-celery-apply-async-with-eta-it-does-the-job-immediately/45084288#45084288) approach seems to deal with the actual time specified and not the async execution....unless I am missing something somewhere?

Reply:
`CELERY_TASK_ALWAYS_EAGER` is used in local development, when a celery process isn't running (I don't see any references to it in `commcare-cloud`).  Basically what it means is "do celery synchronously".  This change attempts to force celery to operate asynchronously via another process, but that's not really compatible with the actual execution environment.

I'd probably just skip the retry and raise the exception if `CELERY_TASK_ALWAYS_EAGER == True`

Reply:
@esoergel  Hmmmm...this is a game changer. I didn't actually know this was the case. Running 

> hasattr(settings, "CELERY_TASK_ALWAYS_EAGER")

on both staging and production yields `False`, so yeah, this is the case it seems.

So I guess my `apply_async`-concerns won't be a problem after all?

Reply:
I removed the settings consideration since it's redundant.